### DISCUSSION TITLE
Upstreaming HeuristicResolver to the clang libraries?

### DISCUSSION THREAD

**Discussion by HighCommander4** [clangd/clangd#1662] [General]
[HeuristicResolver](https://searchfox.org/llvm/rev/3a458256ee22a0e7c31529de42fa6caa263d88fe/clang-tools-extra/clangd/HeuristicResolver.h) is a code component in clangd used to aid code navigation features in templated code by heuristically associating dependent names with the declarations they're likely to target.

It's a standalone component with no dependencies on anything else in clangd.

**I would like to propose upstreaming it to the clang libraries**, meaning having it live in `clang/lib/`, with the public interface exposed in `clang/include/`. (The choice of subdirectory remains to be decided, some possible fits could be `AST/`, `Sema/`, or `Tooling/`.)

Upstreaming it in this way would have the following advantages:

 1. Other code in-tree but outside of clangd could make use of it. For example, it would facilitate fixing https://github.com/clangd/clangd/issues/810 if the code in `SemaCodeComplete.cpp` could make use of `HeuristicResolver`.
 2. Out-of-tree tools could make use of it. Associating dependent names with likely target declarations is something that various code indexer tools want to do as well. One such tool, [Searchfox](https://searchfox.org/), which is clang-based and uses clang's public C++ API, currently maintains a [copy](https://github.com/mozsearch/mozsearch/blob/cb2bd4b8d4330f67006a422a05a774b3864ca7a8/clang-plugin/from-clangd/HeuristicResolver.h) of clangd's HeuristicResolver for its own use. If the component lived in the clang libraries and was exposed as public API, then tools like this could make use of it directly.
 3. It would allow better integration between `HeuristicResolver` and other code in the clang libraries that performs similar tasks. For example, [CXXRecordDecl::lookupDependentName()](https://searchfox.org/llvm/rev/3a458256ee22a0e7c31529de42fa6caa263d88fe/clang/lib/AST/CXXInheritance.cpp#432) performs a task that falls under the umbrella of "heuristic resolution", and `HeuristicResolver` builds on it. In a [recent issue](https://github.com/clangd/clangd/issues/1657#issuecomment-1575460953), I've identified a case where it would be helpful if the implementation of `lookupDependentName()` could call into existing helpers in `HeuristicResolver`. Having the component live in the clang libraries would allow for this.

One potential roadblock to upstreaming `HeuristicResolver` is that the component does not currently have tests that exercise its public interface directly; rather, its test coverage is provided indirectly through tests that exercise the clangd features that use it. I'm happy to rectify this by writing tests that exercise the component directly.

Any feedback on this proposal is appreciated.

**Comment by HighCommander4** [comment]
I realized that Sam actually suggested this upstreaming a while ago, in https://reviews.llvm.org/D92290#2462468:

> reuse: might we also want to use the heuristic resolution for code completion at some point, e.g. LHS of MemberExpr? (We could abstract it via some callbacks, but I also suspect it's not coupled to clangd and could just be lifted up to clang at some point)

I'll take that as a sign that I'm barking up the right tree :)

**Comment by HighCommander4** [comment]
> The choice of subdirectory remains to be decided, some possible fits could be `AST/`, `Sema/`, or `Tooling/`.

One potentially relevant consideration here is that some more advanced heuristics that we may want to implement in `HeuristicResolver` (I have an old WIP patch for one at https://reviews.llvm.org/D93522) would benefit from having access to `Sema`.

If I understand the layering of the clang libraries correctly, that would mean `HeuristicResolver` may need to live in `Sema/` rather than `AST/`.

**Comment by kadircet** [comment]
I think the proposal LG, there are definitely other components in clang ecosystem that can benefit from a unified implementation. Especially the existence of a downstream dependency also backs that theory up.

Some other interesting findings I discovered while taking a look:
- `CXXRecordDecl::lookupDependentName` doesn't really need to be a member and only usages are in libindex, some tidy checks, heuristic resolver itself and semacodecomplete. So I believe we should also make that part of heuristicresolver and change other usages to go through heuristicresolver. Since SemaCodeComplete is also a dependency this points towards lifting into `Sema/`.
- `Sema` depends on `AST`. So if we lift `HeuristicResolver` into `AST` we can't really make use of `Sema` in `HeuristicResolver`.
- `SemaCodeComplete` already implements a set of resolution heuristics (`ConceptInfo`, `getApproximateType`) that might again benefit from living in a central component.

So considering all these I am also in favor of lifting into `Sema/`. But I think we should actually mention these in LLVM discourse as well to see if others have concerns. Can you put an RFC there?
The biggest hesitation I have ATM is, there aren't (m)any public interfaces in `Sema` that provides imprecise/inaccurate information.

@hokein FYI

**Comment by hokein** [comment]
+1, it looks like a reasonable proposal to me (and thanks for all your hard work on improving clangd on dependent code!).

As Kadir mentioned above, since we move it to clang, it would be nice to have input from the Clang community (@AaronBallman), writing an RFC in LLVM discourse seems like a good start.


>> If I understand the layering of the clang libraries correctly, that would mean HeuristicResolver may need to live in Sema/ rather than AST/.
> So considering all these I am also in favor of lifting into Sema/

Yeah, if we're going to use `Sema` object in `HeuristicResolver`, `Sema` is the right place.

> One potentially relevant consideration here is that some more advanced heuristics that we may want to implement in HeuristicResolver (I have an old WIP patch for one at https://reviews.llvm.org/D93522) would benefit from having access to Sema.

This is an interesting idea, but the usage of `Sema` makes me a bit nervous -- IIUC, we use Sema to **perform template instantiations** in order to get more concrete type information. There are some concerns:

- performing template instantiations is **expensive** in clang -- performing each per each selection in clangd could significantly increase latency, which may be unaffordable;
- the side effects of performing template instantiations when the AST built is finished are unclear -- the AST can get mutated, we might see unexpected diagnostic behavior, maybe even result in subtle crashes due to some clang invariant violations;
- most clang tools interact with the clang AST, not the `Sema` object (accessing `Sema` might be hard for some tools);

My main concern is performance, it seems too expensive to use for latency-sensitive features (code-completion, go-to-definition, hover), therefore I'm not sure this is the final approach we should pursue. Perhaps there are other, more affordable options to consider.

While it is always nice to make more dependent cases work, one alternative is to draw a line, and accept that these cases are too hard to support.

**Comment by AaronBallman** [comment]
> As Kadir mentioned above, since we move it to clang, it would be nice to have input from the Clang community (@AaronBallman), writing an RFC in LLVM discourse seems like a good start.

+1 to writing an RFC just to make sure there's general agreement in the community that we'd like to maintain this interface and to get more design perspectives.

I think this could make sense to upstream, but it'd need to have a reason to live in-tree in clang. Use by clangd + other potential uses isn't a reason the functionality needs to move into Clang, but if you offered to implement something in Clang which uses this as part of the upstreaming effort (not necessarily the initial commit), that builds a much stronger case. Replacing existing logic for code completion could be a reasonable use case. It would need some form of test coverage, but unit tests would suffice for the initial commit.

**Comment by HighCommander4** [comment]
Tracking next steps on this in https://github.com/llvm/llvm-project/issues/121310.

**Reply by HighCommander4** [reply]
The requested Discourse RFC is now posted at https://discourse.llvm.org/t/rfc-upstream-heuristicresolver-from-clangd-to-libsema/84004

**Reply by HighCommander4** [reply]
The upstreaming has been completed, so I'm going to close this discussion. Thank you everyone for your input.
